app/permission/foreground/android.go

```diff
@@ -0,0 +1,75 @@
+// SPDX-License-Identifier: Unlicense OR MIT
+
+//+build android
+
+/*
+Package foreground implements permissions to run a foreground service.
+See https://developer.android.com/guide/components/foreground-services.
+
+The following entries will be added to AndroidManifest.xml:
+
+    <uses-permission android:name="android.permission.FOREGROUND_SERVICE"/>
+
+*/
+
+package foreground
+
+import (
+	"gioui.org/app"
+	"git.wow.st/gmp/jni"
+)
+
+const (
+	foregroundService = "org/gioui/GioForegroundService"
+)
+
+// StartForeround notifies the system that the program will perform
+// background work and that it shouldn't be killed. It returns a channel
+// that should be closed when the background work is complete.
+
+// StartForeground is a no-op on Linux, Windows, macOS; Android will
+// display a notification during background work; iOS isn't supported.
+func StartForeground() chan struct{} {
+	closeChan := make(chan struct{})
+
+	// get a handle on the startForeground, stopForeground methods of GioForegroundService
+	// run everything in a goroutine so that start/stop calls are from the same thread
+	go func() {
+		jni.Do(jni.JVMFor(app.JavaVM()), func(env jni.Env) error {
+			appCtx := jni.Object(app.AppContext())
+			classLoader := jni.ClassLoaderFor(env, appCtx)
+			foregroundServiceClass, err := jni.LoadClass(env, classLoader, foregroundService)
+			if err != nil {
+				return err
```

Comment:
Expose errors to the caller and don't return from Start until the foregrounding has successfully begun.

Reply:
initially I wrote it this way but realized that when I started a goroutine to wait on the channel before stopping the service, that the applicaiton would crash - I think because the caller is from another thread? I didn't see anything in the logcat output to hint exactly what was happening - but possibly this can be refactored by getting new handles before calling the stopService method.

Reply:
You don't need to switch threads, just have an additional channel for errors that Start can wait on before returning closeChan.

```
errChan := make(chan error)
go func() {
// Process and send to errChan if failed.
...
errChan <- nil
}()

if err := <-errChan; err != nil {
...
}
``